ompi/mca/op/cuda/op_cuda_impl.cu

```diff
@@ -0,0 +1,950 @@
+        const int index = blockIdx.x * blockDim.x + threadIdx.x;                                    \
+        const int stride = blockDim.x * gridDim.x;                                                  \
+        for (int i = index; i < n; i += stride) {                                                   \
+            inout[i] = fn(inout[i], in[i]);                                                         \
+        }                                                                                           \
+    }                                                                                               \
+    void                                                                                            \
+    ompi_op_cuda_2buff_##name##_##type_name##_submit(const type *in,                                \
+                                              type *inout,                                          \
+                                              int count,                                            \
+                                              int threads_per_block,                                \
+                                              int max_blocks,                                       \
+                                              CUstream stream) {                                    \
+        int threads = min(count, threads_per_block);                                                \
+        int blocks  = min((count + threads-1) / threads, max_blocks);                               \
+        int n = count;                                                                              \
+        CUstream s = stream;                                                                        \
+        ompi_op_cuda_2buff_##name##_##type_name##_kernel<<<blocks, threads, 0, s>>>(in, inout, n);  \
+    }
+
+#define FUNC_FUNC(name, type_name, type) FUNC_FUNC_FN(name, type_name, type, current_func)
+
+#if defined(USE_VECTORS)
+#define VFUNC_FUNC(name, type_name, type, vlen, vfn, fn)                                                    \
+    static __global__ void                                                                                  \
+    ompi_op_cuda_2buff_##name##_##type_name##_kernel_v(const type *__restrict__ in,                         \
+                                                       type *__restrict__ inout, int n) {                   \
+        using vtype = Vec<type, vlen>;                                                                      \
+        constexpr const size_t alignment = sizeof(type)*vlen;                                               \
+        const int index = blockIdx.x * blockDim.x + threadIdx.x;                                            \
+        const int stride = blockDim.x * gridDim.x;                                                          \
+        size_t in_pad = ALIGN_PAD_AMOUNT(in, alignment);                                                    \
```

Comment:
any reason why you align to the read only buffer instead of the read/write ?

Reply:
I had to pick one. The caller will make sure they have the same alignment and use the non-vector functions otherwise: https://github.com/open-mpi/ompi/pull/12569/files#diff-e97bc6b8f6f91b1028b5fc665a5cdc15e914c081e7ca61411ad3c97b8a849169R234